### PR TITLE
Integrated data transformer with `sncast` commands

### PR DESCRIPTION
**Pull request by integraledelebesgue**
**Stack**:
- #2555
- #2548 ⬅
- #2547
- #2546
- #2545

⚠️ *Part of a stack created by [spr](https://github.com/ejoffe/spr). Do not merge manually using the UI - doing so may have unexpected results.*

### CODE REVIEW

File: crates/sncast/src/main.rs [hunk 1]
```diff
@@ -249,9 +250,18 @@ async fn run_async_command(
                 .try_into_fee_settings(&provider, account.block_id())
                 .await?;
 
+            let selector = get_selector_from_name("constructor").unwrap();
+
+            let contract_class = get_contract_class(deploy.class_hash, &provider).await?;
```
Comment (cptartur) [inline crates/sncast/src/main.rs:255]: Below we are adding additional context to this, but here not, why is that?

File: crates/sncast/src/main.rs [hunk 2]
```diff
@@ -284,13 +294,25 @@ async fn run_async_command(
...
 
+            let class_hash = get_class_hash_by_address(&provider, contract_address)
+                .await?
+                .with_context(|| {
+                    format!(
+                    	"Couldn't retrieve class hash of a contract with address {contract_address:#x}"
+                    )
+                })?;
+
+            let contract_class = get_contract_class(class_hash, &provider).await?;
```
Comment (cptartur) [inline crates/sncast/src/main.rs:305]: Is calldata not optional here? What in case of functions that have no arguments, do we event support these?

Can be a separate issue.

File: crates/sncast/src/main.rs [hunk 3]
```diff
@@ -331,9 +353,18 @@ async fn run_async_command(
             let selector = get_selector_from_name(&function)
                 .context("Failed to convert entry point selector to FieldElement")?;
 
+            let class_hash = get_class_hash_by_address(&provider, contract_address)
+                .await
+                .with_context(|| format!("Couldn't retrieve class hash of a contract with address {contract_address:#x}"))?
+                .with_context(|| format!("Couldn't retrieve class hash of a contract with address {contract_address:#x}"))?;
```
Comment (cptartur) [inline crates/sncast/src/main.rs:359]: This is duplicated

File: crates/sncast/src/starknet_commands/call.rs [hunk 4]
```diff
@@ -18,9 +18,9 @@ pub struct Call {
     #[clap(long)]
     pub function: String,
 
-    /// Arguments of the called function (list of hex)
+    /// Arguments of the called function (a list of Cairo-like expressions or values serialized as felts)
```
Comment (cptartur) [inline crates/sncast/src/starknet_commands/call.rs:21]: I don't think `Cairo-like expression` is a term all users would understand. Perhaps this should show an example, be explained or just reworded?

File: crates/sncast/tests/e2e/deploy.rs [hunk 5]
```diff
@@ -308,7 +308,7 @@ fn test_wrong_calldata() {
         output,
         indoc! {r"
         command: deploy
-        error: An error occurred in the called contract[..]Failed to deserialize param #2[..]
+        error: An error occurred in the called contract[..]('Input too long for arguments')[..]
```
Comment (cptartur) [inline crates/sncast/tests/e2e/deploy.rs:311]: Why was this changed?

File: crates/sncast/tests/e2e/deploy.rs [hunk 6]
```diff
@@ -330,14 +330,11 @@ async fn test_contract_not_declared() {
...
-    let output = snapbox.assert().success();
+    let output = snapbox.assert().failure();
 
     assert_stderr_contains(
         output,
-        indoc! {r"
-        command: deploy
-        error: An error occurred in the called contract[..]Class with hash[..]is not declared[..]
-        "},
+        "Error: Couldn't retrieve contract class with hash: 0x1",
```
Comment (cptartur) [inline crates/sncast/tests/e2e/deploy.rs:337]: This is kind of "breaking" in terms of that we are loosing an explanation of why is this contract missing. Previously it said that it is not required. I'm not sure if that's good.

File: crates/sncast/tests/e2e/invoke.rs [hunk 7]
```diff
@@ -310,14 +310,11 @@ async fn test_contract_does_not_exist() {
...
-    let output = snapbox.assert().success();
+    let output = snapbox.assert().failure();
 
     assert_stderr_contains(
         output,
-        indoc! {r"
-        command: invoke
-        error: An error occurred in the called contract[..]Requested contract address[..]is not deployed[..]
-        "},
+        "Error: Couldn't retrieve class hash of a contract with address 0x1",
```
Comment (cptartur) [inline crates/sncast/tests/e2e/invoke.rs:317]: Same here

File: crates/sncast/tests/e2e/main_tests.rs [hunk 8]
```diff
@@ -27,14 +27,11 @@ async fn test_happy_case_from_sncast_config() {
...
-    let output = snapbox.assert().success();
+    let output = snapbox.assert().failure();
 
     assert_stderr_contains(
         output,
-        indoc! {r"
-        command: call
-        error: There is no contract at the specified address
-        "},
+        "Error: Couldn't retrieve class hash of a contract with address 0x0",
```
Comment (cptartur) [inline crates/sncast/tests/e2e/main_tests.rs:34]: And here

File: crates/sncast/src/main.rs (hunk 3 above)
Comment (integraledelebesgue) [reply crates/sncast/src/main.rs:362]: Indeed. `get_class_hash_by_address` returns `Result<Option<Felt>>`. We need to handle both `Err` and `None`

File: crates/sncast/tests/e2e/deploy.rs (hunk 6 above)
Comment (integraledelebesgue) [reply crates/sncast/tests/e2e/deploy.rs]: Made an issue [#2554](https://github.com/foundry-rs/starknet-foundry/issues/2554#issue-2564353877)

**Review by cptartur — COMMENTED** [review]
Don't like we are changing the errors so much. However, I think behavior to previous can still be retained, as I've outlined in the relevant comment.

File: crates/sncast/src/starknet_commands/call.rs (hunk 4 above)
Comment (integraledelebesgue) [reply crates/sncast/src/starknet_commands/call.rs:21]: Sure, I will rewrite it in other words

File: crates/sncast/tests/helpers/constants.rs [hunk 9]
```diff
@@ -10,7 +10,7 @@ pub const NETWORK: &str = "testnet";
 pub const SEED: u32 = 1_053_545_548;
 
 // Block number used by devnet to fork the Sepolia testnet network in the tests
-pub const FORK_BLOCK_NUMBER: u32 = 68_000;
+pub const FORK_BLOCK_NUMBER: u32 = 100_000;
```
[the remark this answers is not in the document]
Comment (integraledelebesgue) [reply crates/sncast/tests/helpers/constants.rs:13]: That's because of the data transformer contract - we need it in our fork of Sepolia. It's been deployed later in the chain so I bumped the block number.

**Comment by integraledelebesgue** [comment]
✓ Commit merged in pull request [#2555](https://github.com/foundry-rs/starknet-foundry/pull/2555)
